File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void solve(vector<int> numbers, int target, int sum, int idx)
{
    if (idx == numbers.size())
    {
        if (sum == target) answer++;
        return;
    }
    solve(numbers, target, sum + numbers[idx], idx + 1);
    solve(numbers, target, sum - numbers[idx], idx + 1);
}

int solution(vector<int> numbers, int target)
{
    solve(numbers, target, 0, 0);
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public int solution(int[] numbers, int target) {
        return this.go(numbers, target);
    }
    
    private int go(int[] numbers, int target) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[] {0, 0});
   
        int count = 0;
        while (!q.isEmpty()) {
            int[] cur = q.poll();

            if (cur[0] == numbers.length) {
                if (cur[1] == target) {
                    count++;
                }
                continue;
            }
            
            q.offer(new int[] {cur[0] + 1, cur[1] + numbers[cur[0]]});
            q.offer(new int[] {cur[0] + 1, cur[1] - numbers[cur[0]]});
        }
 
        return count;
    }
}